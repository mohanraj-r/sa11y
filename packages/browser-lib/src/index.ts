/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export { base, recommended, full } from '@sa11y/preset-rules';
export { exceptionListFilter } from '@sa11y/format';
export { assertAccessible } from '@sa11y/assert';

export const nameSpace = 'sa11y';
