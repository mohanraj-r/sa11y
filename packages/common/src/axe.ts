/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import * as axe from 'axe-core';

export const axeRuntimeExceptionMsgPrefix = 'Error running accessibility checks using axe:';

export const axeVersion: string | undefined = axe.version;

/**
 * Interface that represents a function that runs axe and returns violations
 */
interface AxeRunner {
    (): Promise<axe.Result[]>;
}

/**
 * Get violations by running axe with given function
 * @param axeRunner - function satisfying AxeRunner interface
 */
export async function getViolations(axeRunner: AxeRunner): Promise<axe.Result[]> {
    let violations;
    try {
        violations = await axeRunner();
    } catch (e) {
        throw new Error(`${axeRuntimeExceptionMsgPrefix} ${(e as Error).message}`);
    }
    return violations;
}
