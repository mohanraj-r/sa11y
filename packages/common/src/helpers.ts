/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

/**
 * Convenience wrapper to prefix a standard header for console log messages.
 */
export function log(...args: unknown[]): void {
    console.log('â™¿[Sa11y]', ...args);
}
