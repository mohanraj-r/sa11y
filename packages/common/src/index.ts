/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export { A11yConfig, AxeResults, axeRuntimeExceptionMsgPrefix, axeVersion, getViolations } from './axe';
export { AssertFunction, Options, WDIOBrowser } from './wdio';
export { errMsgHeader, ExceptionList } from './format';
