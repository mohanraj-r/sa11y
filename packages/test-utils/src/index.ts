/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export {
    audioURL,
    domWithA11yIssues,
    domWithA11yIssuesBodyID,
    domWithNoA11yIssues,
    htmlFileWithA11yIssues,
    htmlFileWithNoA11yIssues,
    a11yIssuesCount,
    shadowDomID,
    videoURL,
} from './test-data';
export { beforeEachSetup, cartesianProduct, checkA11yError } from './utils';
