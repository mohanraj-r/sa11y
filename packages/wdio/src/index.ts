/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export { assertAccessible, assertAccessibleSync, Options } from './wdio';
export { setup } from './setup';
