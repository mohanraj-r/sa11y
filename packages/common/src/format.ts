/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { AxeResults } from './axe';

export const errMsgHeader = 'accessibility issues found';

/**
 * Filter to post-process a11y results from axe
 */
export interface Filter {
    (violations: AxeResults, ...args: never[]): AxeResults;
}

// TODO (refactor): constrain rule id to known rule ids e.g using string literal, keyof, in etc
//  e.g. https://stackoverflow.com/a/54061487
// const ruleIDs = getRules().map((ruleObj) => ruleObj.ruleId);
// type RuleID = keyof ruleIDs;
// type RuleID = typeof ruleIDs[number];
type RuleID = string;
type CssSelectors = string[];
/**
 * Exception list of map of rule to corresponding css targets that needs to be filtered from a11y results.
 */
export type ExceptionList = Record<RuleID, CssSelectors>;
