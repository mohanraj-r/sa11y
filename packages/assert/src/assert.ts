/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import * as axe from 'axe-core';
import { A11yConfig, recommended } from '@sa11y/preset-rules';
import { A11yError } from '@sa11y/format';
import { getViolations } from '@sa11y/common';

/**
 * Type def for context that can be checked for accessibility.
 * Limiting to this subset from all options supported by axe for ease of use and maintenance.
 */
export type A11yCheckableContext = Document | HTMLElement;

/**
 * Checks DOM for accessibility issues and throws an error if violations are found.
 * @param context - DOM or HTMLElement to be tested for accessibility
 * @param rules - A11yConfig preset rule to use, defaults to recommended
 * @throws error - with the accessibility issues found, does not return any value
 * */
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export async function assertAccessible(context: A11yCheckableContext = document, rules: A11yConfig = recommended) {
    const violations = await getViolations(async () => {
        const results = await axe.run(context as axe.ElementContext, rules as axe.RunOptions);
        return results.violations;
    });

    A11yError.checkAndThrow(violations);
}
