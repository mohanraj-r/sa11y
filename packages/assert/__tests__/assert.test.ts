/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { assertAccessible } from '../src/assert';
import { recommended, getA11yConfig, base } from '@sa11y/preset-rules';
import {
    beforeEachSetup,
    checkA11yError,
    audioURL,
    videoURL,
    domWithA11yIssues,
    domWithNoA11yIssues,
    shadowDomID,
} from '@sa11y/test-utils';
import { axeRuntimeExceptionMsgPrefix } from '@sa11y/common';

beforeEach(() => {
    beforeEachSetup();
});

/**
 * Test util to test DOM with a11y issues
 */
// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
async function testDOMWithA11yIssues() {
    document.body.innerHTML = domWithA11yIssues;
    expect.assertions(3);
    await assertAccessible(document, recommended).catch((e: Error) => {
        checkA11yError(e);
    });
}

describe('assertAccessible API', () => {
    it('should trigger axe runtime exception for non existent rule', async () => {
        const errConfig = getA11yConfig(['non-existent-rule']);
        expect.assertions(2);
        await assertAccessible(document, errConfig).catch((e: Error) => {
            expect(e).toBeTruthy();
            expect(e.toString()).toContain(axeRuntimeExceptionMsgPrefix);
        });
    });

    // eslint-disable-next-line jest/expect-expect
    it.each([base, recommended])(
        'should throw no errors for dom with no a11y issues with config %#',
        async (config) => {
            document.body.innerHTML = domWithNoA11yIssues;
            await assertAccessible(document, config); // No error thrown
        }
    );

    it.each([
        // DOM to test, expected assertions
        [domWithNoA11yIssues, 0],
        [domWithA11yIssues, 3],
    ])(
        'should use default document, ruleset, formatter when called with no args - expecting %# assertion',
        async (testDOM: string, expectedAssertions: number) => {
            expect.assertions(expectedAssertions);
            document.body.innerHTML = testDOM;
            await assertAccessible().catch((e) => checkA11yError(e));
        }
    );

    // eslint-disable-next-line jest/expect-expect
    it('should throw an error with a11y issues found for dom with a11y issues', testDOMWithA11yIssues);

    it('should not throw error with HTML element with no a11y issues', async () => {
        document.body.innerHTML = domWithNoA11yIssues;
        const elem = document.getElementById(shadowDomID);
        expect(elem).toBeTruthy();
        await assertAccessible(elem); // No error thrown
    });

    it('should throw error with HTML element with a11y issues', async () => {
        expect.assertions(5);
        document.body.innerHTML = domWithA11yIssues;
        const elements = document.getElementsByTagName('body');
        expect(elements).toHaveLength(1);
        const elem = elements[0];
        expect(elem).toBeTruthy();
        await assertAccessible(elem).catch((e) => checkA11yError(e));
    });

    it.each(['', 'non-existent-audio.mp3', audioURL])(
        'should test audio without timing-out using src %#',
        async (source: string) => {
            document.body.innerHTML = `<audio src=${source}>Audio test</audio>`;
            await assertAccessible(document, getA11yConfig(['audio-caption', 'no-autoplay-audio']));
        }
    );

    it.each(['', 'non-existent-video.webm', videoURL])(
        'should test video without timing-out using src %#',
        async (source: string) => {
            document.body.innerHTML = `<video src=${source}>Video test</video>`;
            await assertAccessible(document, getA11yConfig(['video-caption', 'no-autoplay-audio']));
        }
    );
});
