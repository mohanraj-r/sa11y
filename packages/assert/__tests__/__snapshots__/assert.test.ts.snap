// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assertAccessible API should throw an error with a11y issues found for dom with a11y issues 1`] = `
"4 Accessibility issues found
 * (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.2/document-title

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.2/link-name

* (landmark-one-main) Document should have one main landmark: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.2/landmark-one-main

* (page-has-heading-one) Page should contain a level-one heading: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.2/page-has-heading-one"
`;

exports[`assertAccessible API should throw error with HTML element with a11y issues 1`] = `
"1 Accessibility issues found
 * (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.2/link-name"
`;

exports[`assertAccessible API should use default document, ruleset, formatter when called with no args - expecting 1 assertion 1`] = `
"5 Accessibility issues found
 * (bypass) Page must have means to bypass repeated blocks: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.1/bypass

* (document-title) Documents must have <title> element to aid in navigation: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.1/document-title

* (link-name) Links must have discernible text: a
	- Help URL: https://dequeuniversity.com/rules/axe/4.1/link-name

* (landmark-one-main) Document must have one main landmark: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.1/landmark-one-main

* (page-has-heading-one) Page must contain a level-one heading: html
	- Help URL: https://dequeuniversity.com/rules/axe/4.1/page-has-heading-one"
`;
